### ISSUE TITLE
List of Work Items per Pull Request

### ISSUE THREAD

**Issue by striky1** [rfennell/AzurePipelines#861]
Hi,

I'm using Tasks XplatGenerateReleaseNotes@3 and WikiUpdaterTask@1 on Release pipeline.

My question is: How I can generate Release Notes with list of Work Items per PR? 

What I want to achieve per Release:

Release base information:
Release ID, created time and so on..

List of Pull Requests with Work Items:
- PR 1
- - Title/Desc/Time and so on..
- - Work Item 1 (title, author, link)
- - Work Item 2 (title, author, link)
- PR 2
- - Title/Desc/Time and so on..
- - Work Item 3 (title, author, link)
- - Work Item 4 (title, author, link)

I was looking for some examples or issues here but what I found based on your answers is that it's possible to do opposite behaviour:

List of Work Items:
- Work Item 1 (title, author, link)
- - PR 1
- Work Item 2 (title, author, link)
- - PR 1
- Work Item 3 (title, author, link)
- - PR 2
- Work Item 4 (title, author, link)
- - PR 2

Any ideas how to make it? 

Thank you a lot

**Comment by rfennell** [comment]
Without a change to the code this is not possible. 

The array of PRs returned by the Azure DevOps API does no contain list of associated WI, but a URL to call to get the list. So I would need to run this extra query to populate the list. It is possible but a change. It will also need some Handlebars help usage to do some dynamic lookups

An alternative that is currently available is to list the WI with their associated PRs, as detailed here #839, does that help?

**Comment by striky1** [comment]
Well, as I wrote above, I found it before I created this Issue and it's not what we need to have as result for customers/project managers to know who, when and what did on which Pull Request.

Another alternative can be to make custom handler to go through Array of Work Items and make new one paired per Pull Request. What do you think? Am I right?

**Comment by rfennell** [comment]
I think trying to 'reverse generate' the list will be awkward, maybe possible, but much easier to end have the PR list with the WI information.

**Comment by striky1** [comment]
So, what are the final thoughts? Not sure now what I can do :)

**Comment by rfennell** [comment]
I will make the change to the code ASAP and provide a sample

**Comment by rfennell** [comment]
I have enriched the PRs dataset with easy to access to WI ID & URL.

This can be used in a template, it is basic lookup between array and then a bit of URL editing to get to the right format

```
## Associated Pull Requests ({{pullRequests.length}})
{{#forEach pullRequests}}
* **[{{this.pullRequestId}}]({{replace (replace this.url "_apis/git/repositories" "_git") "pullRequests" "pullRequest"}})** {{this.title}}
{{#forEach this.associatedWorkitems}}
   {{#with (lookup_a_work_item ../../relatedWorkItems this.url)}}
    - WI [{{this.id}}]({{replace this.url "_apis/wit/workItems" "_workitems/edit"}}) - {{lookup this.fields 'System.Title'}} 
   {{/with}}
{{/forEach}}
{{/forEach}}
```

This is currently in test and will be released so 3.25.x

**Comment by striky1** [comment]
@rfennell it works! Ty a lot
